client/src/controllers/TabsController.ts

```diff
@@ -0,0 +1,352 @@
+      // Wait for hidden attribute to be applied then fade in
+      setTimeout(() => {
+        tabContent.classList.add(this.animateClass);
+      }, this.transitionValue);
+    }, this.transitionValue);
+  }
+
+  animateOut(tabContent: HTMLElement) {
+    // Wait element to transition out and then hide with hidden
+    tabContent.classList.remove(this.animateClass);
+    setTimeout(() => {
+      // eslint-disable-next-line no-param-reassign
+      tabContent.hidden = true;
+    }, this.transitionValue);
+  }
+
+  hideTabContent(tabId: string) {
+    if (tabId === this.activeValue || !this.activeValue) {
+      return;
+    }
+
+    const tabContent = this.getTabPanelByHref(tabId);
+    if (!tabContent) {
+      return;
+    }
+    if (this.animateValue) {
+      this.animateOut(tabContent);
+    } else {
+      tabContent.hidden = true;
+    }
+
+    const tab = this.getTabLabelByHref(tabId);
+    if (!tab) {
+      return;
+    }
+    tab.setAttribute('aria-selected', 'false');
+    tab.setAttribute('tabindex', '-1');
+  }
+
+  handleKeydown(event: KeyboardEvent) {
+    const keyPressed = event.key;
+
+    switch (keyPressed) {
+      case Keys.Left:
+      case Keys.Right:
+        this.switchTabOnArrowPress(event);
+        break;
+      case Keys.End:
+        event.preventDefault();
+        this.focusLastTab();
+        break;
+      case Keys.Home:
+        event.preventDefault();
+        this.focusFirstTab();
+        break;
+      default:
+        break;
+    }
+  }
+
+  switchTabOnArrowPress(event) {
+    const pressed = event.key;
+    if (!event.target) {
+      return;
+    }
+
+    let direction: number = 0;
+    if (pressed === Keys.Left) {
+      direction = KeyWeight.ArrowLeft;
+    }
+    if (pressed === Keys.Right) {
+      direction = KeyWeight.ArrowRight;
+    }
+
+    const tabIndex = (event.target as IndexedEventTarget).index;
+    const tab = this.labelTargets[tabIndex + direction];
+    if (tabIndex !== undefined) {
+      if (tab) {
+        this.activeValue = tab.getAttribute('aria-controls') as string;
+        tab.focus();
+      } else if (pressed === Keys.Left) {
+        this.focusLastTab();
+      } else if (pressed === Keys.Right) {
+        this.focusFirstTab();
+      }
+    }
+  }
+
+  focusFirstTab() {
+    const tab = this.labelTargets[0];
+    this.activeValue = tab.getAttribute('aria-controls') as string;
+    tab.focus();
+  }
+
+  focusLastTab() {
+    const tab = this.labelTargets[this.labelTargets.length - 1];
+    this.activeValue = tab.getAttribute('aria-controls') as string;
+    tab.focus();
+  }
+
+  loadHistory(event: PopStateEvent) {
+    if (event.state && event.state.tabContent) {
+      const tab = this.getTabLabelByHref(event.state.tabContent);
+      if (tab) {
+        this.activeValue = event.state.tabContent;
+        tab.focus();
+      }
+    }
+  }
+
+  validate() {
+    this.labelTargets.forEach((label, idx) => {
+      const panel = this.panelTargets[idx];
```

Comment:
I would just collect a bunch of errors and then throw them at the end.

So use `map` here instead of `forEach`, generate an array of arrays.

Then at the end throw a `new Error` with that data.

Reply:
should i pass the whole list or iterate over errors and throw error?

Reply:
I think you are right, this is fine for now. We should focus on the main behaviour. We can always enhance later.

Console.warn is good.

Thank you.